File: rustbridge-java/rustbridge-core/src/main/java/com/rustbridge/PluginConfig.java
package com.rustbridge;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.util.HashMap;
import java.util.Map;

/**
 * Configuration for plugin initialization.
 */
public class PluginConfig {
    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper()
            .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);

    private final Map<String, Object> data;
    private Map<String, Object> initParams;
    private Integer workerThreads;
    private String logLevel = "info";
    private int maxConcurrentOps = 1000;
    private long shutdownTimeoutMs = 5000;

    /**
     * Create a new empty configuration.
     */
    public PluginConfig() {
        this.data = new HashMap<>();
    }

    /**
     * Create a configuration with default settings.
     *
     * @return a new configuration with defaults
     */
    public static PluginConfig defaults() {
        return new PluginConfig();
    }

    /**
     * Set the number of worker threads.
     *
     * @param threads the number of threads
     * @return this config for chaining
     */
    public PluginConfig workerThreads(int threads) {
        this.workerThreads = threads;
        return this;
    }

    /**
     * Set the log level.
     *
     * @param level the log level
     * @return this config for chaining
     */
    public PluginConfig logLevel(LogLevel level) {
        this.logLevel = level.name().toLowerCase();
        return this;
    }

    /**
     * Set the log level from a string.
     * <p>
     * Valid values are: "trace", "debug", "info", "warn", "error", "off".
     * Case-insensitive.
     *
     * @param level the log level as a string
     * @return this config for chaining
     */
    public PluginConfig logLevel(String level) {
        this.logLevel = level.toLowerCase();
        return this;
    }

    /**
     * Set the maximum concurrent operations.
     *
     * @param maxOps the maximum concurrent operations
     * @return this config for chaining
     */
    public PluginConfig maxConcurrentOps(int maxOps) {
        this.maxConcurrentOps = maxOps;
        return this;
    }

    /**
     * Set the shutdown timeout.
     *
     * @param timeoutMs the timeout in milliseconds
     * @return this config for chaining
     */
    public PluginConfig shutdownTimeout(long timeoutMs) {
        this.shutdownTimeoutMs = timeoutMs;
        return this;
    }

    /**
     * Set the shutdown timeout in milliseconds.
     * <p>
     * Alias for {@link #shutdownTimeout(long)} to match the field name.
     *
     * @param timeoutMs the timeout in milliseconds
     * @return this config for chaining
     */
    public PluginConfig shutdownTimeoutMs(long timeoutMs) {
        this.shutdownTimeoutMs = timeoutMs;
        return this;
    }

    /**
     * Set a custom configuration value.
     *
     * @param key   the configuration key
     * @param value the configuration value
     * @return this config for chaining
     */
    public PluginConfig set(String key, Object value) {
        this.data.put(key, value);
        return this;
    }

    /**
     * Set an initialization parameter.
     * <p>
     * Initialization parameters are passed to the plugin during startup and are
     * intended for one-time setup configuration (migrations, seed data, etc.).
     * These are separate from runtime configuration in {@code data}.
     *
     * @param key   the parameter key
     * @param value the parameter value
     * @return this config for chaining
     */
    public PluginConfig initParam(String key, Object value) {
        if (this.initParams == null) {
            this.initParams = new HashMap<>();
        }
        this.initParams.put(key, value);
        return this;
    }

    /**
     * Set all initialization parameters at once.
     * <p>
     * Initialization parameters are passed to the plugin during startup and are
     * intended for one-time setup configuration. This replaces any existing init params.
     *
     * @param params map of initialization parameters
     * @return this config for chaining
     */
    public PluginConfig initParams(Map<String, Object> params) {
        this.initParams = new HashMap<>(params);
        return this;
    }

    /**
     * Serialize the configuration to JSON bytes.
     *
     * @return the JSON bytes
     */
    public byte[] toJsonBytes() {
        ObjectNode json = OBJECT_MAPPER.createObjectNode();
        json.set("data", OBJECT_MAPPER.valueToTree(data));

        if (initParams != null && !initParams.isEmpty()) {
            json.set("init_params", OBJECT_MAPPER.valueToTree(initParams));
        }

        if (workerThreads != null) {
            json.put("worker_threads", workerThreads);
        }

        json.put("log_level", logLevel);
        json.put("max_concurrent_ops", maxConcurrentOps);
        json.put("shutdown_timeout_ms", shutdownTimeoutMs);

        try {
            return OBJECT_MAPPER.writeValueAsBytes(json);
        } catch (JsonProcessingException e) {
            throw new RuntimeException("Failed to serialize config", e);
        }
    }
}

File: rustbridge-csharp/RustBridge.Core/PluginConfig.cs
using System.Text.Json;
using System.Text.Json.Nodes;

namespace RustBridge;

/// <summary>
/// Configuration for plugin initialization.
/// </summary>
public class PluginConfig
{
    private readonly Dictionary<string, object?> _data = new();
    private Dictionary<string, object?>? _initParams;
    private int? _workerThreads;
    private string _logLevel = "info";
    private int _maxConcurrentOps = 1000;
    private long _shutdownTimeoutMs = 5000;

    /// <summary>
    /// Create a new empty configuration.
    /// </summary>
    public PluginConfig()
    {
    }

    /// <summary>
    /// Create a configuration with default settings.
    /// </summary>
    /// <returns>A new configuration with defaults.</returns>
    public static PluginConfig Defaults() => new();

    /// <summary>
    /// Set the number of worker threads.
    /// </summary>
    /// <param name="threads">The number of threads.</param>
    /// <returns>This config for chaining.</returns>
    public PluginConfig WorkerThreads(int threads)
    {
        _workerThreads = threads;
        return this;
    }

    /// <summary>
    /// Set the log level.
    /// </summary>
    /// <param name="level">The log level.</param>
    /// <returns>This config for chaining.</returns>
    public PluginConfig WithLogLevel(LogLevel level)
    {
        _logLevel = level.ToString().ToLowerInvariant();
        return this;
    }

    /// <summary>
    /// Set the log level from a string.
    /// <para>
    /// Valid values are: "trace", "debug", "info", "warn", "error", "off".
    /// Case-insensitive.
    /// </para>
    /// </summary>
    /// <param name="level">The log level as a string.</param>
    /// <returns>This config for chaining.</returns>
    public PluginConfig WithLogLevel(string level)
    {
        _logLevel = level.ToLowerInvariant();
        return this;
    }

    /// <summary>
    /// Set the maximum concurrent operations.
    /// </summary>
    /// <param name="maxOps">The maximum concurrent operations.</param>
    /// <returns>This config for chaining.</returns>
    public PluginConfig MaxConcurrentOps(int maxOps)
    {
        _maxConcurrentOps = maxOps;
        return this;
    }

    /// <summary>
    /// Set the shutdown timeout.
    /// </summary>
    /// <param name="timeoutMs">The timeout in milliseconds.</param>
    /// <returns>This config for chaining.</returns>
    public PluginConfig ShutdownTimeout(long timeoutMs)
    {
        _shutdownTimeoutMs = timeoutMs;
        return this;
    }

    /// <summary>
    /// Set the shutdown timeout in milliseconds.
    /// Alias for <see cref="ShutdownTimeout"/> to match the field name.
    /// </summary>
    /// <param name="timeoutMs">The timeout in milliseconds.</param>
    /// <returns>This config for chaining.</returns>
    public PluginConfig ShutdownTimeoutMs(long timeoutMs)
    {
        _shutdownTimeoutMs = timeoutMs;
        return this;
    }

    /// <summary>
    /// Set a custom configuration value.
    /// </summary>
    /// <param name="key">The configuration key.</param>
    /// <param name="value">The configuration value.</param>
    /// <returns>This config for chaining.</returns>
    public PluginConfig Set(string key, object? value)
    {
        _data[key] = value;
        return this;
    }

    /// <summary>
    /// Set an initialization parameter.
    /// <para>
    /// Initialization parameters are passed to the plugin during startup and are
    /// intended for one-time setup configuration (migrations, seed data, etc.).
    /// These are separate from runtime configuration in <c>data</c>.
    /// </para>
    /// </summary>
    /// <param name="key">The parameter key.</param>
    /// <param name="value">The parameter value.</param>
    /// <returns>This config for chaining.</returns>
    public PluginConfig InitParam(string key, object? value)
    {
        _initParams ??= new Dictionary<string, object?>();
        _initParams[key] = value;
        return this;
    }

    /// <summary>
    /// Set all initialization parameters at once.
    /// <para>
    /// Initialization parameters are passed to the plugin during startup and are
    /// intended for one-time setup configuration. This replaces any existing init params.
    /// </para>
    /// </summary>
    /// <param name="parameters">Map of initialization parameters.</param>
    /// <returns>This config for chaining.</returns>
    public PluginConfig InitParams(IDictionary<string, object?> parameters)
    {
        _initParams = new Dictionary<string, object?>(parameters);
        return this;
    }

    /// <summary>
    /// Serialize the configuration to JSON bytes.
    /// </summary>
    /// <returns>The JSON bytes.</returns>
    public byte[] ToJsonBytes()
    {
        var json = new JsonObject
        {
            ["data"] = JsonSerializer.SerializeToNode(_data),
            ["log_level"] = _logLevel,
            ["max_concurrent_ops"] = _maxConcurrentOps,
            ["shutdown_timeout_ms"] = _shutdownTimeoutMs
        };

        if (_initParams is { Count: > 0 })
        {
            json["init_params"] = JsonSerializer.SerializeToNode(_initParams);
        }

        if (_workerThreads.HasValue)
        {
            json["worker_threads"] = _workerThreads.Value;
        }

        return JsonSerializer.SerializeToUtf8Bytes(json);
    }
}
